[bug] Generated CMake files cause errors with `-Werror=dev --warn-uninitialized`

Conan generated CMake files cause errors when the project is configured with `-Werror=dev --warn-uninitialized`.

### Environment Details (include every applicable attribute)
  * Operating System+version: Windows 10
  * Compiler+version: VS2019
  * Conan version: 1.44.0
  * Python version: 3.10.0

### Steps to reproduce (Include if Applicable)

`find_package()` a dependency after installing it with `CMakeToolchain` and `cmake_find_package_multi` generators.

### Logs (Executed commands with output) (Include/Attach if Applicable)

Example log attached for when using the toolchain to `find_package(freetype)`. It's just a lot of uninitialized variable use.

[log.log](https://github.com/conan-io/conan/files/7833584/log.log)

Adding checks (and fixes) for CMakeDeps too.

I am struggling with a couple of vars:

```
         uninitialized variable 'dep_COMPONENT_NAMES'
E           Call Stack (most recent call first):
E             dep-config.cmake:11 (include)
E             CMakeLists.txt:7 (find_package)
E           This error is for project developers. Use -Wno-error=dev to suppress it.
E           
E           -- Conan: Target declared 'dep::dep'
E           CMake Error (dev) at dep-config.cmake:14 (foreach):
E             uninitialized variable 'dep_FIND_DEPENDENCY_NAMES'

```
they seem to be properly defined in the xxx-data.cmake files, but for some reason it keeps complaining.

--- a/conan/tools/cmake/cmakedeps/templates/macros.py
+++ b/conan/tools/cmake/cmakedeps/templates/macros.py
@@ -26,9 +26,9 @@ class MacrosTemplate(CMakeDepsFileTemplate):
     @property
     def template(self):
         return textwrap.dedent("""
-        function(conan_message MESSAGE_OUTPUT)
+        function(conan_message MESSAGE_TYPE MESSAGE_CONTENT)
             if(NOT CONAN_CMAKE_SILENT_OUTPUT)
-                message(${ARGV${0}})
+                message(${MESSAGE_TYPE} ${MESSAGE_CONTENT})
             endif()
         endfunction()
 
@@ -48,7 +48,9 @@ class MacrosTemplate(CMakeDepsFileTemplate):
        endmacro()
 
        function(conan_package_library_targets libraries package_libdir deps out_libraries out_libraries_target config_suffix package_name)
-           unset(_CONAN_ACTUAL_TARGETS CACHE)
+           set(_out_libraries "")
+           set(_out_libraries_target "")
+           set(_CONAN_ACTUAL_TARGETS "")
 
            foreach(_LIBRARY_NAME ${libraries})
                find_library(CONAN_FOUND_LIBRARY NAMES ${_LIBRARY_NAME} PATHS ${package_libdir}
@@ -65,7 +67,7 @@ class MacrosTemplate(CMakeDepsFileTemplate):
                        # Create a micro-target for each lib/a found
                        add_library(${_LIB_NAME} UNKNOWN IMPORTED)
                        set_target_properties(${_LIB_NAME} PROPERTIES IMPORTED_LOCATION ${CONAN_FOUND_LIBRARY})
-                       set(_CONAN_ACTUAL_TARGETS ${_CONAN_ACTUAL_TARGETS} ${_LIB_NAME})
+                       list(APPEND _CONAN_ACTUAL_TARGETS ${_LIB_NAME})
                    else()
                        conan_message(STATUS "Skipping already existing target: ${_LIB_NAME}")
                    endif()

--- a/conan/tools/cmake/cmakedeps/templates/target_data.py
+++ b/conan/tools/cmake/cmakedeps/templates/target_data.py
@@ -59,10 +59,18 @@ class ConfigDataTemplate(CMakeDepsFileTemplate):
               ########### AGGREGATED COMPONENTS AND DEPENDENCIES FOR THE MULTI CONFIG #####################
               #############################################################################################
 
+              {% if components_names %}
               list(APPEND {{ pkg_name }}_COMPONENT_NAMES {{ components_names }})
               list(REMOVE_DUPLICATES {{ pkg_name }}_COMPONENT_NAMES)
+              {% else %}
+              set({{ pkg_name }}_COMPONENT_NAMES "")
+              {% endif %}
+              {% if dependency_filenames %}
               list(APPEND {{ pkg_name }}_FIND_DEPENDENCY_NAMES {{ dependency_filenames }})
               list(REMOVE_DUPLICATES {{ pkg_name }}_FIND_DEPENDENCY_NAMES)
+              {% else %}
+              set({{ pkg_name }}_FIND_DEPENDENCY_NAMES "")
+              {% endif %}
 
               ########### VARIABLES #######################################################################
               #############################################################################################

--- a/conan/tools/cmake/cmakedeps/templates/target_data.py
+++ b/conan/tools/cmake/cmakedeps/templates/target_data.py
@@ -59,9 +59,9 @@ class ConfigDataTemplate(CMakeDepsFileTemplate):
               ########### AGGREGATED COMPONENTS AND DEPENDENCIES FOR THE MULTI CONFIG #####################
               #############################################################################################
 
-              set({{ pkg_name }}_COMPONENT_NAMES {{ '${'+ pkg_name }}_COMPONENT_NAMES} {{ components_names }})
+              list(APPEND {{ pkg_name }}_COMPONENT_NAMES {{ components_names }})
               list(REMOVE_DUPLICATES {{ pkg_name }}_COMPONENT_NAMES)
-              set({{ pkg_name }}_FIND_DEPENDENCY_NAMES {{ '${'+ pkg_name }}_FIND_DEPENDENCY_NAMES} {{ dependency_filenames }})
+              list(APPEND {{ pkg_name }}_FIND_DEPENDENCY_NAMES {{ dependency_filenames }})
               list(REMOVE_DUPLICATES {{ pkg_name }}_FIND_DEPENDENCY_NAMES)
 
               ########### VARIABLES #######################################################################

--- a/conan/tools/cmake/toolchain.py
+++ b/conan/tools/cmake/toolchain.py
@@ -158,7 +158,7 @@ class FPicBlock(Block):
 class GLibCXXBlock(Block):
     template = textwrap.dedent("""
         {% if set_libcxx %}
-        set(CONAN_CXX_FLAGS "${CONAN_CXX_FLAGS} {{ set_libcxx }}")
+        string(APPEND CONAN_CXX_FLAGS " {{ set_libcxx }}")
         {% endif %}
         {% if glibcxx %}
         add_definitions(-D_GLIBCXX_USE_CXX11_ABI={{ glibcxx }})
@@ -215,10 +215,10 @@ class SkipRPath(Block):
 
 class ArchitectureBlock(Block):
     template = textwrap.dedent("""
-        set(CONAN_CXX_FLAGS "${CONAN_CXX_FLAGS} {{ arch_flag }}")
-        set(CONAN_C_FLAGS "${CONAN_C_FLAGS} {{ arch_flag }}")
-        set(CONAN_SHARED_LINKER_FLAGS "${CONAN_SHARED_LINKER_FLAGS} {{ arch_flag }}")
-        set(CONAN_EXE_LINKER_FLAGS "${CONAN_EXE_LINKER_FLAGS} {{ arch_flag }}")
+        string(APPEND CONAN_CXX_FLAGS " {{ arch_flag }}")
+        string(APPEND CONAN_C_FLAGS " {{ arch_flag }}")
+        string(APPEND CONAN_SHARED_LINKER_FLAGS " {{ arch_flag }}")
+        string(APPEND CONAN_EXE_LINKER_FLAGS " {{ arch_flag }}")
         """)
 
     def context(self):
@@ -266,8 +266,8 @@ class SharedLibBock(Block):
 
 class ParallelBlock(Block):
     template = textwrap.dedent("""
-        set(CONAN_CXX_FLAGS "${CONAN_CXX_FLAGS} /MP{{ parallel }}")
-        set(CONAN_C_FLAGS "${CONAN_C_FLAGS} /MP{{ parallel }}")
+        string(APPEND CONAN_CXX_FLAGS " /MP{{ parallel }}")
+        string(APPEND CONAN_C_FLAGS " /MP{{ parallel }}")
         """)
 
     def context(self):
@@ -402,10 +402,10 @@ class FindConfigFiles(Block):
         {% endif %}
         # To support the generators based on find_package()
         {% if cmake_module_path %}
-        set(CMAKE_MODULE_PATH {{ cmake_module_path }} ${CMAKE_MODULE_PATH})
+        list(PREPEND CMAKE_MODULE_PATH {{ cmake_module_path }})
         {% endif %}
         {% if cmake_prefix_path %}
-        set(CMAKE_PREFIX_PATH {{ cmake_prefix_path }} ${CMAKE_PREFIX_PATH})
+        list(PREPEND CMAKE_PREFIX_PATH {{ cmake_prefix_path }})
         {% endif %}
         {% if android_prefix_path %}
         set(CMAKE_FIND_ROOT_PATH {{ android_prefix_path }} ${CMAKE_FIND_ROOT_PATH})
@@ -471,10 +471,18 @@ class UserToolchain(Block):
 
 class CMakeFlagsInitBlock(Block):
     template = textwrap.dedent("""
-        set(CMAKE_CXX_FLAGS_INIT "${CMAKE_CXX_FLAGS_INIT} ${CONAN_CXX_FLAGS}")
-        set(CMAKE_C_FLAGS_INIT "${CMAKE_C_FLAGS_INIT} ${CONAN_C_FLAGS}")
-        set(CMAKE_SHARED_LINKER_FLAGS_INIT "${CMAKE_SHARED_LINKER_FLAGS_INIT} ${CONAN_SHARED_LINKER_FLAGS}")
-        set(CMAKE_EXE_LINKER_FLAGS_INIT "${CMAKE_EXE_LINKER_FLAGS_INIT} ${CONAN_EXE_LINKER_FLAGS}")
+        if(DEFINED CONAN_CXX_FLAGS)
+          string(APPEND CMAKE_CXX_FLAGS_INIT " ${CONAN_CXX_FLAGS}")
+        endif()
+        if(DEFINED CONAN_C_FLAGS)
+          string(APPEND CMAKE_C_FLAGS_INIT " ${CONAN_C_FLAGS}")
+        endif()
+        if(DEFINED CONAN_SHARED_LINKER_FLAGS)
+          string(APPEND CMAKE_SHARED_LINKER_FLAGS_INIT " ${CONAN_SHARED_LINKER_FLAGS}")
+        endif()
+        if(DEFINED CONAN_EXE_LINKER_FLAGS)
+          string(APPEND CMAKE_EXE_LINKER_FLAGS_INIT " ${CONAN_EXE_LINKER_FLAGS}")
+        endif()
         """)
 
 

--- a/conans/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps.py
+++ b/conans/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps.py
@@ -339,4 +339,4 @@ def test_private_transitive():
     client.run("install consumer -g CMakeDeps -s arch=x86_64 -s build_type=Release")
     assert "dep/0.1:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Skip" in client.out
     data_cmake = client.load("pkg-release-x86_64-data.cmake")
-    assert "set(pkg_FIND_DEPENDENCY_NAMES ${pkg_FIND_DEPENDENCY_NAMES} )" in data_cmake
+    assert 'set(pkg_FIND_DEPENDENCY_NAMES "")' in data_cmake

--- a/conans/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_components.py
+++ b/conans/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_components.py
@@ -69,8 +69,7 @@ class PropagateSpecificComponents(unittest.TestCase):
         t.run('install middle/version@ -g CMakeDeps')
 
         content = t.load('middle-release-x86_64-data.cmake')
-        self.assertIn("set(middle_FIND_DEPENDENCY_NAMES ${middle_FIND_DEPENDENCY_NAMES} top)",
-                      content)
+        self.assertIn("list(APPEND middle_FIND_DEPENDENCY_NAMES top)", content)
 
         content = t.load('middle-config.cmake')
         self.assertIn("find_dependency(${_DEPENDENCY} REQUIRED NO_MODULE)", content)

--- a/conans/test/functional/toolchains/cmake/test_cmake_toolchain.py
+++ b/conans/test/functional/toolchains/cmake/test_cmake_toolchain.py
@@ -190,17 +190,22 @@ def test_cmaketoolchain_no_warnings():
         from conans import ConanFile
         class Conan(ConanFile):
             settings = "os", "compiler", "arch", "build_type"
-            generators = "CMakeToolchain"
+            generators = "CMakeToolchain", "CMakeDeps"
+            requires = "dep/0.1"
         """)
     consumer = textwrap.dedent("""
        set(CMAKE_CXX_COMPILER_WORKS 1)
        set(CMAKE_CXX_ABI_COMPILED 1)
        project(MyHello CXX)
        cmake_minimum_required(VERSION 3.15)
+
+       find_package(dep CONFIG REQUIRED)
        """)
-    client.save({"conanfile.py": conanfile,
+    client.save({"dep/conanfile.py": GenConanfile("dep", "0.1"),
+                 "conanfile.py": conanfile,
                  "CMakeLists.txt": consumer})
 
+    client.run("create dep")
     client.run("install .")
     client.run_command("cmake . -DCMAKE_TOOLCHAIN_FILE=./conan_toolchain.cmake "
                        "-Werror=dev --warn-uninitialized")

--- a/conans/test/functional/toolchains/cmake/test_cmake_toolchain.py
+++ b/conans/test/functional/toolchains/cmake/test_cmake_toolchain.py
@@ -177,3 +177,32 @@ def test_cmake_toolchain_multiple_user_toolchain():
     client.run("create . pkg/0.1@")
     assert "mytoolchain1.cmake !!!running!!!" in client.out
     assert "mytoolchain2.cmake !!!running!!!" in client.out
+
+
+@pytest.mark.tool_cmake
+def test_cmaketoolchain_no_warnings():
+    """Make sure unitialized variables do not cause any warnings, passing -Werror=dev
+    and --wanr-unitialized, calling "cmake" with conan_toolchain.cmake used to fail
+    """
+    # Issue https://github.com/conan-io/conan/issues/10288
+    client = TestClient()
+    conanfile = textwrap.dedent("""
+        from conans import ConanFile
+        class Conan(ConanFile):
+            settings = "os", "compiler", "arch", "build_type"
+            generators = "CMakeToolchain"
+        """)
+    consumer = textwrap.dedent("""
+       set(CMAKE_CXX_COMPILER_WORKS 1)
+       set(CMAKE_CXX_ABI_COMPILED 1)
+       project(MyHello CXX)
+       cmake_minimum_required(VERSION 3.15)
+       """)
+    client.save({"conanfile.py": conanfile,
+                 "CMakeLists.txt": consumer})
+
+    client.run("install .")
+    client.run_command("cmake . -DCMAKE_TOOLCHAIN_FILE=./conan_toolchain.cmake "
+                       "-Werror=dev --warn-uninitialized")
+    assert "Using Conan toolchain"
+    # The real test is that there are no errors, it returns successfully

--- a/conans/test/integration/toolchains/cmake/cmakedeps/test_cmakedeps.py
+++ b/conans/test/integration/toolchains/cmake/cmakedeps/test_cmakedeps.py
@@ -37,7 +37,7 @@ def test_package_from_system():
     assert not os.path.exists(os.path.join(client.current_folder, "dep2-config.cmake"))
     assert not os.path.exists(os.path.join(client.current_folder, "custom_dep2-config.cmake"))
     contents = client.load("dep1-release-x86_64-data.cmake")
-    assert 'set(dep1_FIND_DEPENDENCY_NAMES ${dep1_FIND_DEPENDENCY_NAMES} custom_dep2)' in contents
+    assert 'list(APPEND dep1_FIND_DEPENDENCY_NAMES custom_dep2)' in contents
 
 
 def test_test_package():

